File: app/src/main/java/ru/michanic/mymot/Models/Category.java
package ru.michanic.mymot.Models;

import android.os.Build;
import android.util.Log;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.RealmResults;
import io.realm.annotations.PrimaryKey;
import ru.michanic.mymot.Utils.ModelsSortComparator;

public class Category extends RealmObject {

    @PrimaryKey
    private int id;

    private String description;
    private String code;
    private String image;
    private String name;
    private int sort;

    // связи
    private RealmList<Model> models;

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getImage() {
        return image;
    }

    public void setModels(RealmList<Model> models) {
        this.models = models;
    }

    public RealmList<Model> getModels() {
        return models;
    }

    public List<Model> getManufacturerModels(int manufacturerId) {
        List<Model> modelsList = new ArrayList<>();
        for (Model model: getModels()) {
            if (model.getManufacturer().getId() == manufacturerId) {
                modelsList.add(model);
            }
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Collections.sort(modelsList, new ModelsSortComparator());
        }
        return modelsList;
    }

}

File: app/src/main/java/ru/michanic/mymot/Kotlin/Models/Category.kt
package ru.michanic.mymot.Kotlin.Models

import android.os.Build
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import ru.michanic.mymot.Kotlin.Utils.ModelsSortComparator
import java.util.*

open class Category : RealmObject() {
    @PrimaryKey
    var id: Int = 0
    private var description: String? = null
    private var code: String? = null
    var image: String? = null
    var name: String? = null
    private var sort: Int = 0

    var models: RealmList<Model>? = null
    fun getManufacturerModels(manufacturerId: Int): List<Model> {
        val modelsList: MutableList<Model> = ArrayList()
        for (model in models ?: emptyList()) {
            if (model.manufacturer?.id == manufacturerId) {
                modelsList.add(model)
            }
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            Collections.sort(modelsList, ModelsSortComparator())
        }
        return modelsList
    }
}